examples/fluids/src/setupstg_shur14.c

```diff
@@ -0,0 +1,333 @@
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Implementation of the Synthetic Turbulence Generation (STG) algorithm
+/// presented in Shur et al. 2014
+
+#include <stdlib.h>
+#include <math.h>
+#include <petsc.h>
+#include "../qfunctions/stg_shur14.h"
+#include "setupstg_shur14.h"
+
+#ifndef M_PI
+#define M_PI    3.14159265358979323846
+#endif
+
+#ifndef NOMINMAX
+
+#ifndef max
+#define max(a,b)            (((a) > (b)) ? (a) : (b))
+#endif
+
+#ifndef min
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+#endif
+
+#endif  /* NOMINMAX */
+
+/*
+ * @brief Perform Cholesky decomposition on array of symmetric 3x3 matrices
+ *
+ * This assumes the input matrices are in order [11,22,33,12,13,23]. This
+ * format is also used for the output.
+ *
+ * @param[in] nprofs Number of matrices in Rij
+ * @param[in] Rij Array of the symmetric matrices [6,nprofs]
+ * @param[out] Cij Array of the Cholesky Decomposition matrices, [6,nprofs]
+ */
+static inline void CalcCholeskyDecomp(int nprofs,
+                                      const CeedScalar Rij[6][nprofs], CeedScalar Cij[6][nprofs]) {
+
+  CeedPragmaSIMD
+  for(int i=0; i<nprofs; i++) {
+    Cij[0][i] = sqrt(Rij[0][i]);
+    Cij[3][i] = Rij[3][i] / Cij[0][i];
+    Cij[1][i] = sqrt(Rij[1][i] - pow(Cij[3][i], 2) );
+    Cij[4][i] = Rij[4][i] / Cij[0][i];
+    Cij[5][i] = (Rij[5][i] - Cij[3][i]*Cij[4][i]) / Cij[1][i];
+    Cij[2][i] = sqrt(Rij[2][i] - pow(Cij[4][i], 2) - pow(Cij[5][i], 2));
+  }
+}
+
+
+/*
+ * @brief Get the number of rows for the PHASTA file at path
+ *
+ * Assumes that the first line of the file has the number of rows and columns
+ * as the only two entries, separated by a single space
+ *
+ * @param[in] path Path to the file
+ * @param[out] nrows Number of rows
+ */
+static inline PetscErrorCode GetNRows(const char path[PETSC_MAX_PATH_LEN],
+                                      int *nrows) {
+
+  PetscErrorCode ierr;
+  int ndims;
+  FILE *fp;
+  const int char_array_len = 512;
+  char line[char_array_len];
+  MPI_Comm comm = PETSC_COMM_WORLD;
+  char **array;
+
+  PetscFunctionBeginUser;
+  ierr = PetscFOpen(comm, path, "r", &fp); CHKERRQ(ierr);
+
+  if ( fgets(line, char_array_len, fp) != NULL) {
```

Comment:
You can use `PetscSynchronizedFGets` for the reads, which allows all ranks to pass through the same code (and then you don't have to broadcast later). It does the broadcast internally, which is fine for a file of this size (relatively small number of lines). If not, then best not to use collective `PetscFOpen` above.

Reply:
Ah. The documentation for `PetscFOpen` says only the first rank actually opens the file, while `PetscSynchronizedFGets` says all the processors get it. I had interpreted that as  `PetscSynchronizedFGets` has all the ranks read the same file at once, but synchronizes the line reading. Hindsight, it makes sense that a HPC library wouldn't include a function that does something like that.

Reply:
Right, nothing slows a supercomputer to a crawl like the filesystem metadata impact of every rank opening files (the same or different files). The bcasts are in the tens of microseconds range and thus no big deal to read once that way until you have many thousands of lines. It's somewhat faster to read larger blocks and broadcast, or create the whole data structure and broadcast, but sometimes that code gets out of sync with a changing file format and it's generally not worth the human time for smallish files.